---
title: "Cichlid OR sc-Seq"
author: "Theresa Alexander reesyxan@umd.edu"
date: "`r Sys.Date()`"
output:
  html_document:
    code_download: true
    code_folding: show
    fig_caption: true
    fig_height: 7
    fig_width: 7
    highlight: tango
    keep_md: false
    mode: selfcontained
    number_sections: true
    self_contained: true
    theme: readable
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
  rmdformats::readthedown:
    code_download: true
    code_folding: show
    df_print: paged
    fig_caption: true
    fig_height: 7
    fig_width: 7
    highlight: tango
    width: 300
    keep_md: false
    mode: selfcontained
    toc_float: true
  BiocStyle::html_document:
    code_download: true
    code_folding: show
    fig_caption: true
    fig_height: 7
    fig_width: 7
    highlight: tango
    keep_md: false
    mode: selfcontained
    toc_float: true
---
<style type="text/css">
body, td {
  font-size: 16px;
}
code.r{
  font-size: 16px;
}
pre {
 font-size: 16px
}
</style>


# Setup 
```{r, message = FALSE}
library(dplyr)
library(Seurat)
library(scuttle)
library(patchwork)
library(miQC)
library(stringr)
library(hpgltools)
library(xlsx)
library(biomaRt)
library(diem)
library(gprofiler2)
library(readr)
library(clustree)
```

## Additional Functions
```{r}
#function to add a hgnc id column to DE results
add_hnci_ids <- function(DE_df, convert_df) {
  split_ids <- str_split(DE_df$intersection, ",")
  gene_ids <- list()
  for (i in 1:length(split_ids)) {
    gene_ids[[i]] <- convert_df %>%
        filter(target %in% split_ids[[i]]) %>%
      select(Row.names) %>%
      as.list()
    DE_df[i, "gene_ids"] <- paste0(gene_ids[[i]]$Row.names, collapse = ", ")
  }
  return(DE_df)
}

GO_genes_avg_expression <- function(seur_obj, clusters, features, slot = "counts"){
  mean_exprs <- data.frame(row.names = features)
  #split seurat object into cluster
  for (i in clusters){
  seur_clust <- subset(x = seur_obj, idents = i)
  #get counts
  counts <- as.data.frame(GetAssayData(object = seur_clust, slot = slot))
  #subset out features OI
  counts_oi <- counts[features,]
  mean_exp <- rowMeans(counts_oi)
  colname <- paste0("cluster", i, "_mean_expr")
  mean_exprs[[colname]] <- mean_exp
  colnamemax <- paste0("cluster", i, "_max_expr")
  mean_exprs[[colnamemax]] <- apply(counts_oi, 1, max)
  }
  return(mean_exprs)
}
```

Info for use if needed:

Fish Clade annotation info
  https://useast.ensembl.org/info/genome/genebuild/2018_12_fish_clade_gene_annotation.pdf
  
  <br>
  
# Data

## CellRanger Counts Matrix

```{r}
# Load the cichlid OE dataset
juntti.data.OE1 <- Read10X(data.dir = "../OE1_data/processed_reads")
juntti.data.OE2 <- Read10X(data.dir = "../OE2_data/processed_reads")

# Initialize the Seurat object
juntti_OE1 <- CreateSeuratObject(counts = juntti.data.OE1, project = "juntti_2trt_2ctrl", 
                                 min.cells = 3, min.features = 200)
juntti_OE2 <- CreateSeuratObject(counts = juntti.data.OE2, project = "juntti_2trt_2ctrl", 
                                 min.cells = 3, min.features = 200)
```

```{r, echo=FALSE, eval=FALSE}
#gff <- load_gff_annotations("~/scratch/reesyxan/juntti_cichlid/reference/Astatotilapia_burtoni_genomeNCBI2/genes/genes_mod.gtf")
#gtf <- read.table(file = "~/scratch/reesyxan/juntti_cichlid/reference/Astatotilapia_burtoni_genomeNCBI2/genes/genes_mod.gtf", 
#                header = FALSE, 
#                sep = '\t')
#colnames(gtf) <- c("seqname", "source", "feature", "start", 
#                   "end", "score", "strand", "frame", "attribute")

#gtf_txs <- gtf[gtf$feature == "transcript",]
#gtf_txs <- gtf_txs[order(gtf_txs$seqname, gtf_txs$start),]

#descgene <- c()
#for (i in 1:length(all.genes)) {
#    att <- gtf_txs[grep(all.genes[i], gtf_txs$attribute)[1], "attribute"]
#    att <- str_split(att, ";") %>%
#      unlist()
#    desc <- unlist(str_split(att[grep("product", att)], pattern = "product"))[2]
#    if (!is.null(desc)){
#    gene_desc <- paste0(all.genes[i], ":", desc)
#    descgene <- c(descgene, gene_desc)
#    } else {
#      descgene <- c(descgene, all.genes[i])
#    }
#}

#write.table(descgene, file = "~/scratch/reesyxan/juntti_cichlid/juntti_git/geneID.description.csv")
```

## GFF Annotation Table
```{r}
#gff <- load_gff_annotations("~/scratch/reesyxan/juntti_cichlid/reference/Astatotilapia_burtoni_genomeNCBI2/genes/genes_mod.gtf")

genedesc <- read.table("~/scratch/reesyxan/juntti_cichlid/juntti_git/geneID.description.csv")
```

<br><br>

https://www.ncbi.nlm.nih.gov/nuccore/NC_027289.1 <- MT IDs. count these versus # MT genes below
<br>
## List of OR Genes of interest

```{r}
OR_genes <- read.xlsx2("~/scratch/reesyxan/juntti_cichlid/data/OR_genes.xlsx", sheetIndex = 1)
```


## Read in output from souporcell

```{r, message = FALSE}
soc_clusters.OE1 <- read_tsv("~/scratch/reesyxan/juntti_cichlid/OE1_data/soup_out_4k/soup_out/clusters.tsv", col_names = TRUE)
soc_clusters.OE2 <- read_tsv("~/scratch/reesyxan/juntti_cichlid/OE2_data/souporcell/OE2_souporcell/clusters.tsv", col_names = TRUE)

dim(soc_clusters.OE1[!is.na(soc_clusters.OE1$barcode),])
dim(juntti_OE1)


cells.use <- head(colnames(pbmc_small))
subset_pbmc <- subset(pbmc_small, cells = cells.use)



soc_clusters.OE1 <- soc_clusters.OE1[soc_clusters.OE1$barcode %in% Cells(juntti_OE1),]
soc_clusters.OE2 <- soc_clusters.OE2[soc_clusters.OE2$barcode %in% Cells(juntti_OE2),]

soc_clusters.OE1 <- as.data.frame(soc_clusters.OE1[match(soc_clusters.OE1$barcode, Cells(juntti_OE1)),])
soc_clusters.OE2 <- as.data.frame(soc_clusters.OE2[match(soc_clusters.OE2$barcode, Cells(juntti_OE2)),])


rownames(soc_clusters.OE2) <- soc_clusters.OE2$barcode

juntti_OE1 <- AddMetaData(object = juntti_OE1, metadata = soc_clusters.OE1$status, col.name = "souporcell_status")
juntti_OE1 <- AddMetaData(object = juntti_OE1, metadata = soc_clusters.OE1$assignment, col.name = "souporcell_assignment")

juntti_OE2 <- AddMetaData(object = juntti_OE2, metadata = soc_clusters.OE2$status, col.name = "souporcell_status")
juntti_OE2 <- AddMetaData(object = juntti_OE2, metadata = soc_clusters.OE2$assignment, col.name = "souporcell_assignment")


OE.combined <- merge(juntti_OE1, y = juntti_OE2, add.cell.ids = c("OE1", "OE2"))
OE.combined

all.genes <- rownames(OE.combined)

```

These are the metrics for how many cells were assigned to "singlet", "doublet", or "unassigned"
```{r}
table(juntti_OE$souporcell_status)
```

I was (falsely) under the impression that Cell Ranger automatically detects and filters doublets. This is not the case (and explicitly said on their site here: https://kb.10xgenomics.com/hc/en-us/articles/360005165411-Are-there-methods-for-identifying-multiplets-)
So I will trust the doublet calls from Souporcell and exclude everything either called as a doublet or unassigned. 



```{r}
juntti_OE <- subset(juntti_OE, subset = souporcell_status == "singlet")
```

## Data Cleaning

### Initial QC Metrics

```{r}
# check the data for mitochondrial content ~ dying cells
juntti_OE[["percent.mt"]] <- PercentageFeatureSet(juntti_OE, pattern = "^ACI43-")
```


```{r}
# Visualize QC metrics as a violin plot
VlnPlot(juntti_OE, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

<br>


```{r, echo = FALSE}
message("We usually filter for cells which have MT percent <5%, so we would filter ", round(sum(juntti_OE[["percent.mt"]] > 5)/ nrow(juntti_OE[["percent.mt"]]) * 100, 2), "% of the cells.")
```


```{r}
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.
plot1 <- FeatureScatter(juntti_OE, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2a <- FeatureScatter(juntti_OE, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot3a <- FeatureScatter(juntti_OE, feature1 = "nFeature_RNA", feature2 = "percent.mt")

plot2a 
plot1 
plot3a
```

### Alternative Methods for QC

Cutoff parameters we are going to use going forward:
min_feature<-200
max_feature<-6000
mt_percent<-5

dim_clusters<-13
res_clusters<-0.5

<br>

```{r}
juntti_OE <- subset(juntti_OE, subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & percent.mt < 5)
```


### Normalize and Find Highly Variable Genes

We will log normalize, center and scale, and do a variance stabilizing transformation (VST) 
```{r, message = FALSE}
juntti_OE <- NormalizeData(juntti_OE, normalization.method = "LogNormalize", scale.factor = 10000)
juntti_OE <- ScaleData(juntti_OE, features = all.genes)
juntti_OE <- FindVariableFeatures(juntti_OE, selection.method = "vst", nfeatures = 2000)
```


```{r, warning=FALSE}
# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(juntti_OE), 10)
#top10_desc <- filter(genedesc, grepl(paste(top10, collapse="|"), x))

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(juntti_OE)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)

plot2
```


# Dimensionality Reduction
## PCA

```{r}
juntti_OE <- RunPCA(juntti_OE, features = VariableFeatures(object = juntti_OE))
```

```{r}
#print(juntti_OE[["pca"]], dims = 1:5, nfeatures = 5)
DimPlot(juntti_OE, reduction = "pca")
```

```{r,message=FALSE, warning=FALSE}
DimHeatmap(juntti_OE, dims = 1:10, cells = 500, balanced = TRUE)
ElbowPlot(juntti_OE)
```

# Cluster into Putative Cell Types

First, we create a SNN graph based on the top PCs, then we subset that SNN graph to minimize modularity (based on the louvain community detection algorithm) which yields clusters of related cells. 

I will cluster at several resolutions (lower resolutions = fewer times the SNN will be divided, higher resolution, more clusters). We can then build a tree to visualize how the cells get divided into different clusters as a function of resolution. 

```{r, warning=FALSE, message=FALSE}
juntti_OE <- FindNeighbors(juntti_OE, dims = 1:13)
juntti_OE <- FindClusters(juntti_OE, 
                          resolution = c(0.5))
juntti_OE <- FindClusters(juntti_OE, 
                          resolution = c(0, 0.1, 0.2, 0.3, 0.4, 0.6, 0.7, 0.8, 0.9, 1.0))

juntti_OE <- SetIdent(juntti_OE, value = "RNA_snn_res.0.5")
clustree(juntti_OE)
```

## UMAP/tSNE

```{r, warning=FALSE}
juntti_OE <- RunUMAP(juntti_OE, dims = 1:13)
juntti_OE <- RunTSNE(juntti_OE, dims = 1:13)

DimPlot(juntti_OE, reduction = "umap")
```


## Souporcell Clusters
<br>
```{r}
DimPlot(juntti_OE, reduction = "umap", group.by = "souporcell_assignment")
```


# Marker Comparison for All Cell Types

## Find Markers
The FindAllMarkers() function does a DE comparison between each cluster against all other clusters. Here I am including all putative cell type clusters. 

```{r}
#juntti_OE.markers <- FindAllMarkers(juntti_OE, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
#write.csv(juntti_OE.markers, file = "../data/markers.csv")
juntti_OE.markers <- read.csv("../data/markers.csv")
juntti_OE.markers %>%
  group_by(cluster) %>%
  slice_max(n = 3, order_by = avg_log2FC) %>%
  print.data.frame()
```



## GSEA with Top Cluster Markers

First I will split top DE markers by cluster, convert entrez gene ID to ensembl IDs, then input those markers to gprofiler's gost tool for GSEA. 


```{r}
#get top markers for each cluster
clusters <- paste0(rep("clust_",12), seq(0, 11, by = 1))
clust_topmarkers <- list()
for (i in clusters){
  clust_topmarkers[[i]] <- juntti_OE.markers %>%
    filter(cluster == str_remove(i, "clust_"))
}
```


```{r, warning = FALSE}
clust_markers_ensembl <- list()
gost_all_res <- list()
j = 0
for (i in clust_topmarkers){
  name <- paste0("clust_", j)
  clust_markers_ensembl[[name]] <- gconvert(query = i[["gene"]], 
                                            organism = "hburtoni", 
                                            target="ENSG", 
                                            mthreshold = Inf, 
                                            filter_na = TRUE)
  gost_all_res[[name]] <- gost(query =  clust_markers_ensembl[[name]][["target"]], 
                               organism = "hburtoni",
                               significant = FALSE)
  j = j + 1
}
```
These results have been written to the file `data/GOstRes`

## GOST plots

A few plots just to show what the results look like.

```{r, echo = FALSE}
gost_all_res$clust_0$plot <- gostplot(gost_all_res$clust_0, capped = FALSE, interactive = FALSE)
gost_all_res$clust_1$plot  <- gostplot(gost_all_res$clust_1, capped = FALSE, interactive = FALSE)
gost_all_res$clust_2$plot  <- gostplot(gost_all_res$clust_2, capped = FALSE, interactive = FALSE)
gost_all_res$clust_3$plot  <- gostplot(gost_all_res$clust_3, capped = FALSE, interactive = FALSE)
gost_all_res$clust_4$plot  <- gostplot(gost_all_res$clust_4, capped = FALSE, interactive = FALSE)
gost_all_res$clust_5$plot  <- gostplot(gost_all_res$clust_5, capped = FALSE, interactive = FALSE)
gost_all_res$clust_6$plot  <- gostplot(gost_all_res$clust_6, capped = FALSE, interactive = FALSE)
gost_all_res$clust_7$plot  <- gostplot(gost_all_res$clust_7, capped = FALSE, interactive = FALSE)
gost_all_res$clust_8$plot  <- gostplot(gost_all_res$clust_8, capped = FALSE, interactive = FALSE)
gost_all_res$clust_9$plot  <- gostplot(gost_all_res$clust_9, capped = FALSE, interactive = FALSE)
gost_all_res$clust_10$plot  <- gostplot(gost_all_res$clust_10, capped = FALSE, interactive = FALSE)
gost_all_res$clust_11$plot  <- gostplot(gost_all_res$clust_11, capped = FALSE, interactive = FALSE)
```


These are the GSEA plots for the 3 neuronal clusters.
```{r}
gost_all_res$clust_0$plot
gost_all_res$clust_1$plot
gost_all_res$clust_4$plot
```


# Comparing the Neuronal Cell Clusters

There are 3 GO terms we are interested to look at. These GO terms are for different ion channels. We would like to know if there are genes associated with different ion channels which are represented at higher rates in only some neuronal clusters. To do this we need to do the following:

1. Determine which clusters are neuronal clusters we want to look at
2. Do a DE between just those clusters
3. redo gost with gprofiler to do an unbiased check for these GO terms in the set of DE genes.
4. At this point, we can do some more biased things if we don't see ion channels DE b/t these clusters, but I am hoping the above will yield results we are interested in.


<br>

After speaking with Jessica, I will be using clusters 0, 1, and 4 as the neuronal clusters.

## DE between just these clusters

```{r}
cluster0_neuronal_markers <- FindMarkers(juntti_OE, ident.1 = 0, ident.2 = c(1, 4), min.pct = 0.25, only.pos = TRUE)
cluster1_neuronal_markers <- FindMarkers(juntti_OE, ident.1 = 1, ident.2 = c(0, 4), min.pct = 0.25, only.pos = TRUE)
cluster4_neuronal_markers <- FindMarkers(juntti_OE, ident.1 = 4, ident.2 = c(0, 1), min.pct = 0.25, only.pos = TRUE)

head(cluster0_neuronal_markers, n = 5)
head(cluster1_neuronal_markers, n = 5)
head(cluster4_neuronal_markers, n = 5)
```

## Convert hgnc Gene IDs to ensembl IDs

```{r}
cluster0_neuronal_markers_merged <- merge(cluster0_neuronal_markers, 
                                          gconvert(query = rownames(cluster0_neuronal_markers), 
                                                   organism = "hburtoni", 
                                                   target="ENSG", 
                                                   mthreshold = Inf, 
                                                   filter_na = TRUE),
                                          by.x = 0, by.y = "input") 

cluster1_neuronal_markers_merged <- merge(cluster1_neuronal_markers,  
                                          gconvert(query = rownames(cluster1_neuronal_markers), 
                                                   organism = "hburtoni", 
                                                   target="ENSG", 
                                                   mthreshold = Inf, 
                                                   filter_na = TRUE), 
                                          by.x = 0, by.y = "input")

cluster4_neuronal_markers_merged <- merge(cluster4_neuronal_markers, 
                                          gconvert(query = rownames(cluster4_neuronal_markers), 
                                                   organism = "hburtoni", 
                                                   target="ENSG", 
                                                   mthreshold = Inf, 
                                                   filter_na = TRUE), 
                                          by.x = 0, by.y = "input")

```

 
## GSEA with these markers for each cluster

```{r}
gostres0_neuronal <- gost(query = cluster0_neuronal_markers_merged$target, 
                          organism = "hburtoni", 
                          significant = FALSE)
gostres1_neuronal <- gost(query = cluster1_neuronal_markers_merged$target, 
                          organism = "hburtoni", 
                          significant = FALSE)
gostres4_neuronal <- gost(query = cluster4_neuronal_markers_merged$target, 
                          organism = "hburtoni", 
                          significant = FALSE)
```


```{r, echo=FALSE, eval=FALSE}
#gostres0_neuronal <- simple_gprofiler2(species = "hburtoni", 
#                                       significant = TRUE,
#                                       sig_genes = cluster0_neuronal_markers_merged$target,
#                                       excel = "../excel/gprofiler_neuronal_0.xlsx")
#gostres1_neuronal <- simple_gprofiler2(species = "hburtoni", 
#                                       significant = TRUE,
#                                       sig_genes = cluster1_neuronal_markers_merged$target, 
#                                       excel = "../excel/gprofiler_neuronal_1.xlsx")
#gostres4_neuronal <- simple_gprofiler2(species = "hburtoni", 
#                                       significant = TRUE,
#                                       sig_genes = cluster4_neuronal_markers_merged$target, 
#                                       excel = "../excel/gprofiler_neuronal_4.xlsx")
```


```{r, echo = FALSE, eval=FALSE}
#gostres0_neuronal_GO <- add_hnci_ids(DE_df = gostres0_neuronal$GO, convert_df = cluster0_neuronal_markers_merged)  
#gostres1_neuronal_GO <- add_hnci_ids(DE_df = gostres1_neuronal$GO, convert_df = cluster1_neuronal_markers_merged)  
#gostres5_neuronal_GO <- add_hnci_ids(DE_df = gostres5_neuronal$GO, convert_df = cluster5_neuronal_markers_merged)  
```


## Pulling out GO_ids of interest specifically

The last two GO terms in the list (G-protein activity and cyclic nucleotide-gated ion channel activity) do not have any annotated genes associated with them in the GO database for our Burtoni. 

```{r}
go_terms_query <- c("GO:0005216", "GO:0005261")
ensembl <- useMart("ensembl", dataset="hburtoni_gene_ensembl")
go_terms <- getBM(attributes=c('hgnc_symbol', 'hgnc_id', 'ensembl_gene_id', 'go_id', 'external_gene_name'),
                   filters = 'go', values = go_terms_query, mart = ensembl)

GO0005261 <- go_terms %>%
  filter(go_id %in% go_terms_query) %>%
  filter(go_id == "GO:0005261") 

GO0005216 <- go_terms %>%
  filter(go_id %in% go_terms_query) %>%
  filter(go_id == "GO:0005216")

message("There are ", nrow(GO0005261), " genes in the GO0005261 (cation channel activity) term for H. Burtoni.")
message("There are ", nrow(GO0005216), " genes in the GO0005216 (ion channel activity) term for H. Burtoni.")
```

## Heatmaps for these GO term-associated genes

Subset our expression data for only these genes
```{r, warning=FALSE}
#keep only neuronal cells
juntti_OE_neuronal <- subset(x = juntti_OE, idents = c(0, 1, 4))

GO0005261 <- GO0005261$external_gene_name[GO0005261$external_gene_name %in% rownames(juntti_OE_neuronal)]
GO0005216 <- GO0005216$external_gene_name[GO0005216$external_gene_name %in% rownames(juntti_OE_neuronal)]

juntti_OE_neuronal_counts <- as.data.frame(GetAssayData(object = juntti_OE_neuronal, slot = "counts"))

#heatmap of these channel marker genes
DoHeatmap(object = juntti_OE_neuronal, features = GO0005261, disp.min = 0)
DoHeatmap(object = juntti_OE_neuronal, features = GO0005216, disp.min = 0)
```

## Average expression of each gene per neuronal cell type

```{r}
juntti_OE_neuronal_0 <- subset(x = juntti_OE, idents = c(0))
juntti_OE_neuronal_1 <- subset(x = juntti_OE, idents = c(1))
#juntti_OE_neuronal_4 <- subset(x = juntti_OE, idents = c(4))


GO0005216_avg_exp <- GO_genes_avg_expression(seur_obj = juntti_OE, clusters = c(0,1), features = GO0005216, slot = "scale.data")
GO0005261_avg_exp <- GO_genes_avg_expression(seur_obj = juntti_OE, clusters = c(0,1), features = GO0005261, slot = "scale.data")

write.xlsx2(GO0005261_avg_exp, file = "GOterms_ofinterest_gene_scaled_expression.xlsx", sheetName = "GO0005261")
write.xlsx2(GO0005216_avg_exp, file = "GOterms_ofinterest_gene_scaled_expression.xlsx", sheetName = "GO0005216", append = TRUE)
```

## Dotplots for these genes 
```{r}
DotPlot(juntti_OE_neuronal, features = GO0005261) + RotatedAxis()
DotPlot(juntti_OE_neuronal, features = GO0005216, col.min = 0) + RotatedAxis()
```

```{r}
juntti_OE_neuronal <- RunUMAP(juntti_OE_neuronal, dims = 1:13)
FeaturePlot(juntti_OE_neuronal, features = c("nalcn", "trpm2", "mcoln1b"))
DimPlot(juntti_OE, reduction = "umap")
DimPlot(juntti_OE_neuronal, reduction = "umap")
```

## Subset neuronal clusters based on sensory receptor expression.

How many OR genes from the list are represented in our dataset?
```{r}
message("Of the 175 genes given as chemosensory genes, ", sum(OR_genes$LOC %in% rownames(juntti_OE_neuronal)), " of them are represented in our dataset.")
```

First, a quick visualization to see if any of these ORs are heavily represented in a particular cluster over the others. 
```{r,warning=FALSE, eval=FALSE, echo=FALSE}
DoHeatmap(object = juntti_OE_neuronal, features = OR_genes$LOC[1:20])
DoHeatmap(object = juntti_OE_neuronal, features = OR_genes$LOC[21:40])
DoHeatmap(object = juntti_OE_neuronal, features = OR_genes$LOC[41:60])
DoHeatmap(object = juntti_OE_neuronal, features = OR_genes$LOC[61:80])
DoHeatmap(object = juntti_OE_neuronal, features = OR_genes$LOC[81:100])
DoHeatmap(object = juntti_OE_neuronal, features = OR_genes$LOC[101:140])
DoHeatmap(object = juntti_OE_neuronal, features = OR_genes$LOC[141:175])
#"LOC102296764"
```



Now we want to know how many ORs a single cell expresses.

In this process, I will also remove that one ubiquitous receptor, "LOC102296764".
<br>

```{r}
#grab count data, keep only genes which we have defined as OR genes, and then count the number of non-zero occurrences for each cell
juntti_OE_neuronal_counts <- as.data.frame(GetAssayData(object = juntti_OE_neuronal, slot = "counts"))
CRs_to_keep <- OR_genes$LOC[OR_genes$LOC != "LOC102296764"]
juntti_OR_neuronal <- juntti_OE_neuronal_counts[rownames(juntti_OE_neuronal_counts) %in% CRs_to_keep, ]

CR_df <- data.frame(matrix(nrow = 0,ncol = 4))
colnames(CR_df) <- c("0", "1", "gt_1", "cutoff")
for (i in 1:250){
  juntti_OR_counts <- apply(juntti_OR_neuronal,2,function(x) sum(x > i))
  tbl <- table(juntti_OR_counts)
  tbl <- c(round(tbl[1]/5262,2)*100, round(tbl[2]/5262, 2)*100, gt_1 = round(sum(tbl[3:length(tbl)])/5262, 2)*100, cutoff = i)
  CR_df[i,] <- tbl
}
CR_df[is.na(CR_df)] <- 0

colors <- c("0" = "grey", "1" = "blue", ">1" = "lightblue")

ggplot(CR_df, aes(x = cutoff, y = `0`)) + 
  geom_point(aes(color = "0")) + 
  geom_line(aes(color= "0")) +
  geom_point(aes(x = cutoff, y = `1`, color = "1")) + 
  geom_line(aes(x = cutoff, y = `1`, color = "1")) +
  geom_point(aes(x = cutoff, y = gt_1, color = ">1")) + 
  geom_line(aes(x = cutoff, y = gt_1, color = ">1")) +
  xlim(c(0,40)) + 
  ylab("Percent of OSN's") +
  xlab("Threshold (# of UMIs)") +
  theme_classic() + 
  labs(color = "# of CRs") +
  scale_color_manual(values = colors)
```

We have decided based on this plot to just pick a threshold of 1 UMI to count as "expression" for that CR.
<br>

Based on that criteria, here is a UMAP for how many CR's a given neuronal cell has expressed.
  
  
```{r}
juntti_OR_counts <- apply(juntti_OR_neuronal,2,function(x) sum(x > 1))  

#add this as metadata (first make sure the annotations are in the correct order with match(), then use Seurat's AddMetaData())
OR_counts <- as.integer(juntti_OR_counts[match(names(juntti_OR_counts), Cells(juntti_OE_neuronal))])
juntti_OE_neuronal <- AddMetaData(object = juntti_OE_neuronal, metadata = OR_counts, col.name = "OR_counts")
```

```{r}
DimPlot(juntti_OE_neuronal, reduction = "umap", group.by = "OR_counts")
```
Let's take out the 0 OR containing cells to get a clearer picture of where the multiple-OR cells lie on this UMAP.

```{r}
gt0_OR_cells <- names(juntti_OR_counts[juntti_OR_counts >0])
DimPlot(juntti_OE_neuronal, reduction = "umap", group.by = "OR_counts", cells = gt0_OR_cells)
```

Based on this, it does not appear that the presumed "immature" neuronal cluster in the middle has a larger proportion of high count ORs.

These are the actual counts of OR's per cell per cluster.

```{r}
df <- as.data.frame(cbind("OR_counts" = juntti_OE_neuronal$OR_counts, "seurat_cluster" = juntti_OE_neuronal$RNA_snn_res.0.5))

OR_sums <- as.data.frame(rbind(table(df), "Total Cells" = colSums(table(df))))
rownames(OR_sums) <- c("0 ORs", "1 OR", "2 ORs", "3 ORs", "4 ORs", "Total Cells")
colnames(OR_sums) <- c("Cluster 0", "Cluster 1", "Cluster 4")
OR_sums
```


Of the cells with 2 OR's expressed in them, the majority of them (based on just un-normalized magnitude), are in cluster 0 (which is also the largest cluster).


```{r}
DimPlot(juntti_OE_neuronal, reduction = "umap", group.by = "OR_counts", cells = juntti_OE_neuronal$RNA_snn_res.0.5 == 0) + ggtitle("OR_Counts Cluster 0")
DimPlot(juntti_OE_neuronal, reduction = "umap", group.by = "OR_counts", cells = juntti_OE_neuronal$RNA_snn_res.0.5 == 1) + ggtitle("OR_Counts Cluster 1")
DimPlot(juntti_OE_neuronal, reduction = "umap", group.by = "OR_counts", cells =  juntti_OE_neuronal$RNA_snn_res.0.5 == 4) + ggtitle("OR_Counts Cluster 4")
```


# Coexpression with cfos

Email note:
One note on searching for correlations with the gene fosab. I note that this gene is actually expressed at a higher level in some non-neuronal cells than in the neurons. As a result, markers for those cell types could fall out as the most highly correlated to fosab. Would you modify the search for its most correlated genes by only focusing on those barcodes that fall into the mature neuron clusters? That is, clusters 0 and 1.


```{r}
juntti_OE_0_1 <- subset(x = juntti_OE, idents = c(0, 1))
counts <- as.data.frame(GetAssayData(object = juntti_OE_0_1, slot = "data"))
t_counts <- as.data.frame(t(counts))

#fosab_cor <- cor(t_counts[,-which(colnames(t_counts) %in% c("fosab"))], t_counts$fosab)


matrix_mod<-as.matrix(counts)
gene<-as.numeric(matrix_mod["fosab",])
#correlations<-apply(matrix_mod,1,function(x){cor(gene,x)})

hist(gene)
plot(density(gene))

FeaturePlot(juntti_OE_0_1, features = c("fosab"))
```


# Creating Final Seurat Object to Share

I want to make a clean version of this object which includes, seurat cluster assignments, souporcell cluster assignments, number of ORs per cell, and souporcell singlet/doublet assignment.

```{r}
juntti_OE_new <- CreateSeuratObject(counts = juntti.data, project = "juntti_2trt_2ctrl", 
                                 min.cells = 3, min.features = 200)

juntti_OE_new <- AddMetaData(object = juntti_OE_new, metadata = soc_clusters$status, col.name = "souporcell_status")
juntti_OE_new <- AddMetaData(object = juntti_OE_new, metadata = soc_clusters$assignment, col.name = "souporcell_assignment")
juntti_OE_new <- AddMetaData(juntti_OE_new, juntti_OE_neuronal$OR_counts, "OR_counts")

juntti_OE_new <- AddMetaData(juntti_OE_new, juntti_OE$RNA_snn_res.0, "RNA_snn_res.0")
juntti_OE_new <- AddMetaData(juntti_OE_new, juntti_OE$RNA_snn_res.0.1, "RNA_snn_res.0.1")
juntti_OE_new <- AddMetaData(juntti_OE_new, juntti_OE$RNA_snn_res.0.2, "RNA_snn_res.0.2")
juntti_OE_new <- AddMetaData(juntti_OE_new, juntti_OE$RNA_snn_res.0.3, "RNA_snn_res.0.3")
juntti_OE_new <- AddMetaData(juntti_OE_new, juntti_OE$RNA_snn_res.0.4, "RNA_snn_res.0.4")
juntti_OE_new <- AddMetaData(juntti_OE_new, juntti_OE$RNA_snn_res.0.5, "RNA_snn_res.0.5")
juntti_OE_new <- AddMetaData(juntti_OE_new, juntti_OE$RNA_snn_res.0.6, "RNA_snn_res.0.6")
juntti_OE_new <- AddMetaData(juntti_OE_new, juntti_OE$RNA_snn_res.0.7, "RNA_snn_res.0.7")
juntti_OE_new <- AddMetaData(juntti_OE_new, juntti_OE$RNA_snn_res.0.8, "RNA_snn_res.0.8")
juntti_OE_new <- AddMetaData(juntti_OE_new, juntti_OE$RNA_snn_res.0.9, "RNA_snn_res.0.9")
juntti_OE_new <- AddMetaData(juntti_OE_new, juntti_OE$RNA_snn_res.1, "RNA_snn_res.1")
juntti_OE_new <- AddMetaData(juntti_OE_new, juntti_OE$nCount_RNA, "nCount_RNA")
juntti_OE_new <- AddMetaData(juntti_OE_new, juntti_OE$nFeature_RNA, "nFeature_RNA")
juntti_OE_new <- AddMetaData(juntti_OE_new, juntti_OE$souporcell_status, "souporcell_status")
juntti_OE_new <- AddMetaData(juntti_OE_new, juntti_OE$souporcell_assignment, "souporcell_assignment")
juntti_OE_new <- AddMetaData(juntti_OE_new, juntti_OE$percent.mt, "percent.mt")
juntti_OE_new <- AddMetaData(juntti_OE_new, juntti_OE$seurat_clusters, "seurat_clusters")

#save(juntti_OE_new, file = "seurat_juntti_v2242022.rda")
```



```{r, echo=FALSE, eval=FALSE}
#saveRDS(juntti_OE, file = "../data/juntti_index.rds")
#write.csv(juntti_OE.markers, file = "../data/markers.csv")

write.xlsx2(gostres0$result, file = "../data/GOstRes.xlsx", sheetName = "cluster0")
write.xlsx2(gostres1$result, file = "../data/GOstRes.xlsx", sheetName = "cluster1", append = TRUE)
write.xlsx2(gostres2$result, file = "../data/GOstRes.xlsx", sheetName = "cluster2", append = TRUE)
write.xlsx2(gostres3$result, file = "../data/GOstRes.xlsx", sheetName = "cluster3", append = TRUE)
write.xlsx2(gostres4$result, file = "../data/GOstRes.xlsx", sheetName = "cluster4", append = TRUE)
write.xlsx2(gostres5$result, file = "../data/GOstRes.xlsx", sheetName = "cluster5", append = TRUE)
write.xlsx2(gostres6$result, file = "../data/GOstRes.xlsx", sheetName = "cluster6", append = TRUE)
write.xlsx2(gostres7$result, file = "../data/GOstRes.xlsx", sheetName = "cluster7", append = TRUE)
write.xlsx2(gostres8$result, file = "../data/GOstRes.xlsx", sheetName = "cluster8", append = TRUE)
write.xlsx2(gostres9$result, file = "../data/GOstRes.xlsx", sheetName = "cluster9", append = TRUE)
write.xlsx2(gostres10$result, file = "../data/GOstRes.xlsx", sheetName = "cluster10", append = TRUE)
write.xlsx2(gostres11$result, file = "../data/GOstRes.xlsx", sheetName = "cluster11", append = TRUE)

#write.xlsx2(cluster0_neuronal_markers_merged, file = "../excel/DE_neuronalclusters.xlsx", sheetName = "cluster0")
#write.xlsx2(cluster1_neuronal_markers_merged, file = "../excel/DE_neuronalclusters.xlsx", sheetName = "cluster1", append = TRUE)
#write.xlsx2(cluster4_neuronal_markers_merged, file = "../excel/DE_neuronalclusters.xlsx", sheetName = "cluster4", append = TRUE)

#write.xlsx2(gostres0_neuronal_GO, file = "../excel/gprofiler_neuronalclusters.xlsx", sheetName = "cluster0")
#write.xlsx2(gostres1_neuronal_GO, file = "../excel/gprofiler_neuronalclusters.xlsx", sheetName = "cluster1", append = TRUE)
#write.xlsx2(gostres4_neuronal_GO, file = "../excel/gprofiler_neuronalclusters.xlsx", sheetName = "cluster4", append = TRUE)
```

